src/xsdba/utils.py

```diff
@@ -0,0 +1,1238 @@
+      these limits are linearly interpolated along the month dimension.
+    """
+    dim = group.dim
+    prop = group.prop
+
+    if prop == "group":
+        if "group" in xq.dims:
+            xq = xq.squeeze("group", drop=True)
+        if "group" in yq.dims:
+            yq = yq.squeeze("group", drop=True)
+
+        out = xr.apply_ufunc(
+            _interp_on_quantiles_1D,
+            newx,
+            xq,
+            yq,
+            kwargs={"method": method, "extrap": extrapolation},
+            input_core_dims=[[dim], ["quantiles"], ["quantiles"]],
+            output_core_dims=[[dim]],
+            vectorize=True,
+            dask="parallelized",
+            output_dtypes=[yq.dtype],
+        )
+        return out
+
+    if prop not in xq.dims:
+        xq = xq.expand_dims({prop: group.get_coordinate()})
+    if prop not in yq.dims:
+        yq = yq.expand_dims({prop: group.get_coordinate()})
+
+    # Adding the cyclic bounds fails for string coordinates like seasons
+    # That's why we map the seasons to integers
+    if prop == "season":
+        xq = xq.assign_coords(season=map_season_to_int(xq.season))
+        yq = yq.assign_coords(season=map_season_to_int(yq.season))
+
+    xq = add_cyclic_bounds(xq, prop, cyclic_coords=False)
+    yq = add_cyclic_bounds(yq, prop, cyclic_coords=False)
+    newg = group.get_index(newx, interp=method != "nearest")
+    oldg = xq[prop].expand_dims(quantiles=xq.coords["quantiles"])
+
+    return xr.apply_ufunc(
+        _interp_on_quantiles_2d,
+        newx,
+        newg,
+        xq,
+        yq,
+        oldg,
+        kwargs={"method": method, "extrap": extrapolation},
+        input_core_dims=[
+            [dim],
+            [dim],
+            [prop, "quantiles"],
+            [prop, "quantiles"],
+            [prop, "quantiles"],
+        ],
+        output_core_dims=[[dim]],
+        vectorize=True,
+        dask="parallelized",
+        output_dtypes=[yq.dtype],
+    )
+
+
+def rank(
+    da: xr.DataArray, dim: str | list[str] = "time", pct: bool = False
+) -> xr.DataArray:
+    """Rank data along a dimension.
+
+    Replicates `xr.DataArray.rank` but as a function usable in a Grouper.apply(). Xarray's docstring is below:
+
+    Equal values are assigned a rank that is the average of the ranks that would have been otherwise assigned to all the
+    values within that set. Ranks begin at 1, not 0. If pct, computes percentage ranks, ranging from 0 to 1.
+
+    A list of dimensions can be provided and the ranks are then computed separately for each dimension.
+
+    Parameters
+    ----------
+    da: xr.DataArray
+      Source array.
+    dim : str | list[str], hashable
+      Dimension(s) over which to compute rank.
+    pct : bool, optional
+      If True, compute percentage ranks, otherwise compute integer ranks.
+      Percentage ranks range from 0 to 1, in opposition to xarray's implementation,
+      where they range from 1/N to 1.
```

Comment:
```suggestion
    da : xr.DataArray
        Source array.
    dim : str | list[str], hashable
        Dimension(s) over which to compute rank.
    pct : bool, optional
        If True, compute percentage ranks, otherwise compute integer ranks.
        Percentage ranks range from 0 to 1, in opposition to xarray's implementation,
        where they range from 1/N to 1.
```